src/marqo/vespa/vespa_client.py

```diff
@@ -0,0 +1,318 @@
+        """
+        Create a VespaClient object.
+        Args:
+            config_url: Vespa Deploy API base URL
+            document_url: Vespa Document API base URL
+            query_url: Vespa Query API base URL
+            pool_size: Number of connections to keep in the connection pool
+        """
+        self.config_url = config_url.strip('/')
+        self.document_url = document_url.strip('/')
+        self.query_url = query_url.strip('/')
+        self.http_client = httpx.Client(
+            limits=httpx.Limits(max_keepalive_connections=pool_size, max_connections=pool_size)
+        )
+
+    def close(self):
+        """
+        Close the VespaClient object.
+        """
+        self.http_client.close()
+
+    def deploy_application(self, application: str):
+        """
+        Deploy a Vespa application.
+        Args:
+            application: Path to the Vespa application root directory
+        """
+        endpoint = f'{self.config_url}/application/v2/tenant/default/prepareandactivate'
+
+        gzip_stream = self._gzip_compress(application)
+
+        response = httpx.post(
+            endpoint,
+            headers={'Content-Type': 'application/x-gzip'},
+            data=gzip_stream.read()
+        )
+
+        self._raise_for_status(response)
+
+    def download_application(self) -> str:
+        session_id = self._create_deploy_session()
+        return self._download_application(session_id)
+
+    def query(self, yql: str, hits: int = 10, ranking: str = None, model_restrict: str = None,
+              query_features: Dict[str, Any] = None, **kwargs) -> QueryResult:
+        """
+        Query Vespa.
+        Args:
+            yql: YQL query
+            hits: Number of hits to return
+            ranking: Ranking profile to use
+            model_restrict: Schema to restrict the query to
+            query_features: Query features
+            **kwargs: Additional query parameters
+        Returns:
+            Query result as a VespaQueryResult object
+        """
+        query_features_list = {
+            f'input.query({key})': value for key, value in query_features.items()
+        } if query_features else {}
+        query = {
+            'yql': yql,
+            'hits': hits,
+            'ranking': ranking,
+            'model.restrict': model_restrict,
+            **query_features_list,
+            **kwargs
+        }
+        query = {key: value for key, value in query.items() if value}
+
+        logger.debug(f'Query: {query}')
+
+        try:
+            resp = self.http_client.post(f'{self.query_url}/search/', data=query)
+        except httpx.HTTPError as e:
+            raise VespaError(e) from e
+
+        self._raise_for_status(resp)
+
+        return QueryResult(**resp.json())
+
+    def feed_batch(self,
+                   batch: List[VespaDocument],
+                   schema: str,
+                   concurrency: int = 100,
```

Comment:
is there a risk of adding excess network pressure? especially if end users make many concurrent add_docs calls

Reply:
There is and at some point Vespa starts rejecting connections. Though less likely once you add inference to the mix. I've reduced the default to 10. We'll have to fine-tune this based on the Vespa cluster and also Marqo inference cluster size.